### DISCUSSION TITLE
#[perseus::engine_main] and #[perseus::browser_main]

### DISCUSSION THREAD

**Discussion by ghost** [framesurge/perseus#219] [Q&A]
How to write code under #[perseus::engine_main] and #[perseus::browser_main]?

**Comment by arctic-hen7** [comment]
You've stumbled across a currently undocumented feature, which will be properly documented before v0.4.x is released! In general, you shouldn't use these, since this is a *very* advanced feature that you shouldn't need unless you have a very advanced use case for tinkering with the deep internals of Perseus.

That said, if you really want to play with this, `#[browser_main]` is the entrypoint to your app in the browser, and should call `perseus::run_client(fn)`, where `fn` is the function that returns your `PerseusApp`. As for `#[engine_main]`, that's the entrypoint to your app on the engine (i.e. the server-side, and at build-time), the contents of which usually look like this:

```rust
// Get the operation the user wishes to run from environment variables
let op = perseus::engine::get_op().unwrap();
// Run the default engine with `fn` being the same as it was for the browser main, and `server_fn` being a function that returns your server (see `examples/core/custom_server` for what this could be, it's also the argument to `#[perseus::main(server_fn)]` when you use that)
let exit_code = perseus::engine::run_dflt_engine(op, fn, server_fn).await;
// Terminate the process with whatever exit code we got, so that the user gets some sensible error messages if required
std::process::exit(exit_code);
```

Note that you *must* have both if you choose to use these!

**Reply by ghost** [reply]
I wrote these code:
mod error_pages;
mod templates;
use perseus::{Html, PerseusApp, run_client, engine::{get_op, run_dflt_engine}};

#[perseus::browser_main]
pub fn main<G: Html>() -> PerseusApp<G> {
    PerseusApp::new()
        .template(crate::templates::index::get_template)
        .template(crate::templates::about::get_template)
        .error_pages(crate::error_pages::get_error_pages)
}

#[perseus::engine_main]
let op = get_op().unwrap();
let exit_code = run_dflt_engine(op, fn, perseus_warp::dflt_server).await;

The compiler said:
error: expected item after attributes
  --> src/main.rs:13:1
   |
13 | #[perseus::engine_main]
   | ^^^^^^^^^^^^^^^^^^^^^^^

**Reply by lukechu10** [reply]
The answer is in the compiler error. You need to put the `perseus::engine_main` on top of a function otherwise it doesn't make any sense.

**Reply by ghost** [reply]
> put the `perseus::engine_main` on top of a function

What function?

**Reply by arctic-hen7** [reply]
@auts2022 all the code underneath `#[engine_main]` needs to be inside a function.

**Reply by ghost** [reply]
I didn't find browser_main and engine_main in examples/core/custom_server .

**Reply by ghost** [reply]
mod error_pages;
mod templates;
use perseus::{Html, PerseusApp, run_client, engine::{get_op, run_dflt_engine}};

#[perseus::browser_main]
pub fn main<G: Html>() -> PerseusApp<G> {
PerseusApp::new()
.template(crate::templates::index::get_template)
.template(crate::templates::about::get_template)
.error_pages(crate::error_pages::get_error_pages)
}

#[perseus::engine_main]
pub fn main<G: Html>() -> PerseusApp<G> {
let op = get_op().unwrap();
let exit_code = run_dflt_engine(op, fn, perseus_warp::dflt_server).await;
std::process::exit(exit_code);
}

The compiler said:
error: expected expression, found keyword `fn`
  --> src/main.rs:16:37
   |
16 | let exit_code = run_dflt_engine(op, fn, perseus_warp::dflt_server).await;
   |                                     ^^ expected expression

error: expected expression
  --> src/main.rs:16:37
   |
16 | let exit_code = run_dflt_engine(op, fn, perseus_warp::dflt_server).await;
   |                                     ^^

error[E0432]: unresolved import `perseus::engine`
 --> src/main.rs:3:45
  |
3 | use perseus::{Html, PerseusApp, run_client, engine::{get_op, run_dflt_engine}};
  |                                             ^^^^^^ could not find `engine` in `perseus`

error[E0412]: cannot find type `G` in this scope
 --> src/main.rs:6:38
  |
6 | pub fn main<G: Html>() -> PerseusApp<G> {
  |                                      ^ not found in this scope

**Reply by ghost** [reply]
If I copy-and-paste [examples/core/custom_server/src/main.rs](https://github.com/framesurge/perseus/blob/main/examples/core/custom_server/src/main.rs) and run "perseus snoop serve" and connect to 127.0.0.1:8080, I saw error message:
An error with HTTP code 500 occurred at '/': 'Error: asset './dist/static/xx-XX-index.html' couldn't be read from store Caused by: No such file or directory (os error 2)'.

However, if I run "perseus serve -w" instead of "perseus snoop serve", the website works fine.

**Reply by arctic-hen7** [reply]
Sorry, @auts2022, what does the custom server example have to do with this? I'm just saying that that example shows you what kind of thing could go in the section I mentioned in my original post. You don't need anything else from it.

May I ask why exactly you want to use these macros? Chances are, you don't actually need them. I'd suggest waiting until they're properly documented.

**Reply by ghost** [reply]
Is it possible to write an alternative to GitHub without #[perseus::engine_main] and #[perseus::browser_main]?

**Reply by lukechu10** [reply]
> Is it possible to write an alternative to GitHub without #[perseus::engine_main] and #[perseus::browser_main]?

I'm sorry but you're being quite vague here. We'll need some more concrete details in order to help you.

**Reply by ghost** [reply]
Is it possible to send arbitrary data from users' browsers to engine-side, then the engine-side save that data to hard disk, without using #[perseus::engine_main] and #[perseus::browser_main]?

**Reply by lukechu10** [reply]
> Is it possible to send arbitrary data from users' browsers to engine-side, then the engine-side save that data to hard disk, without using #[perseus::engine_main] and #[perseus::browser_main]?

Yes definitely. As @arctic-hen7 said, these two attributes are very niche and you will probably never need to use them. just look at the custom server example where you can set up API routes just like a regular backend server.

**Reply by arctic-hen7** [reply]
Also @auts2022 remember to run `perseus snoop build` before `perseus snoop serve`, as documented, since snooping bypasses the regular build process for debugging.

**Reply by ghost** [reply]
Can we use #[cfg(not(target_arch = "wasm32"))] and #[cfg(target_arch = "wasm32")] instead of #[perseus::engine_main] and #[perseus::browser_main]?

**Reply by arctic-hen7** [reply]
@auts2022 no, those are completely different. I would strongly suggest you read the Rust book, since Perseus assumes a fairly robust knowledge of the language. The `cfg` macro is designed to only compile the annotated code on a certain target, while the `main` attributes are from Perseus, and define the entrypoints to your app.
